ENH: Implement the NumPy C SIMD vectorization interface

seiko2plus opened the pull request:
### This pullrequest changes

---
#### Implement the NumPy C SIMD vectorization interface

"NPYV" or universal intrinsics as NEP-38 define it, are types and functions
 intended to simplify the vectorization of code on different platforms.

The current implementations support  SIMD extensions SSE, AVX2, AVX512,
 VSX and NEON on the top of C definitions that defined by the new
 generated header '_cpu_dispatch.h' which included by 'cpu_dispatch.h'.

And covers the following operations:

 - memory load and store
 - zero, setall, set, select, reinterpret
 - boolean conversions
 - (add, sub, mul, div, adds, subs)
 - logical
 - comparison
 - left and right shifting
 - combine, zip

Qiyu8 commented on numpy/core/src/common/simd/neon/misc.h:
```diff
@@ -0,0 +1,255 @@
+#define npyv_zero_s32() vdupq_n_s32((int)0)
+#define npyv_zero_u64() vreinterpretq_u64_s32(npyv_zero_s32())
+#define npyv_zero_s64() vreinterpretq_s64_s32(npyv_zero_s32())
+#define npyv_zero_f32() vreinterpretq_f32_s32(npyv_zero_s32())
```
can `vdupq_n_f32(0.0f)` replace `vreinterpretq_f32_s32(npyv_zero_s32())`?

seiko2plus replied:
sure

seiko2plus committed:
ENH: [1/4] implement the NumPy C SIMD vectorization interface

  "NPYV" or universal intrinsics as NEP-38 define it, are types and functions
  intended to simplify vectorization of code on different platforms.

  This patch initialize NPYV for SIMD extensions SSE, AVX2, AVX512,
  VSX and NEON on the top of C definitions that defined by the new
  generated header '_cpu_dispatch.h' which included by 'cpu_dispatch.h'.

seiko2plus committed:
ENH: [2/4] implement the NumPy C SIMD vectorization interface

  implement the following intrinsics for X86 extensions:

    - load, store
    - zero, setall, set, select, reinterpret
    - boolean conversions
    - (add, sub, mul, div, adds, subs)
    - logical
    - comparison
    - left and right shifting
    - combine, zip

seiko2plus committed:
ENH: [3/4] implement the NumPy C SIMD vectorization interface

  implement the same intrinsics as X86 for NEON

seiko2plus committed:
ENH: [4/4] implement the NumPy C SIMD vectorization interface

  implement the same intrinsics as X86 for Power/VSX little-endian mode

seiko2plus commented:
@mattip, Should I move the SIMD module and testing unit into a separated pr?

mattip commented:
> Should I move the SIMD module and testing unit into a separated pr?

Yes please.

mattip commented on numpy/core/include/numpy/npy_common.h:
```diff
@@ -141,6 +141,14 @@
+#elif defined(__GNUC__)
+    #define NPY_FINLINE static NPY_INLINE __attribute__((always_inline))
+#else
+    #define NPY_INLINE static
```
```suggestion
    #define NPY_FINLINE static
```

seiko2plus commented:
@mattip, done. I opened a new pr(#16782) contains the SIMD module and testing units

mattip reviewed (commented):
There is a lot here, very impressive. Is this all new code or was it adapted from somewhere else?

Is there a script we can run on each of the directories to verify that all the universal intrinsics are defined for all the variations?

mattip commented on numpy/core/src/common/simd/simd.h:
```diff
@@ -0,0 +1,56 @@
+#ifdef NPY_HAVE_NEON
+    #include "neon/neon.h"
+#endif
+
```
```suggestion
/* Fallback if none of the above have been defined. Disables SIMD features */
```

seiko2plus replied:
in the extension header, e.g. `avx512/avx512.h`.

seiko2plus commented:
@mattip,
There is a lot here, very impressive. Is this all new code or was it adapted from somewhere else?

No, I wrote it from scratch. I tried to put all my experience on it which mostly I gain it during working on OpenCV.

seiko2plus commented:
@mattip,
Is there a script we can run on each of the directories to verify that all the universal intrinsics are defined for all the variations?

I'm not sure what you actually mean, but I guess yes, the SIMD module and the testing unit check `test_simd.py` in #16782

seiko2plus committed:
SEQ([1/4]): Fix NPY_FINLINE failback

Co-authored-by: Matti Picus <matti.picus@gmail.com>

mattip commented:
> the SIMD module and the testing unit check test_simd.py in #16782

Thanks. LGTM.
